File: Rust/src/two_sum.rs
use std::collections::HashMap;

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut hash_map = HashMap::<i32, i32>::new();
    for (i, v) in nums.iter().enumerate() {
        let complement = target - v;
        if let Some(&value) = hash_map.get(&complement) {
            return vec![value, i as i32];
        } else {
            hash_map.insert(*v, i as i32);
        }
    }
    vec![-1]
}

pub fn two_sum2(nums: Vec<i32>, target: i32) -> Vec<i32> {
    for (i, v) in nums.iter().enumerate() {
        let complement = target - v;
        for (j, n) in nums.iter().enumerate().skip(i + 1) {
            if complement == *n && i != j {
                return vec![i as i32, j as i32];
            }
        }
    }
    vec![-1]
}

File: Ruby/lib/two_sum.rb
# frozen_string_literal: true

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  map = {}
  nums.each_with_index do |num, i|
    complement = target - num
    return [map[complement], i] if map.key?(complement)

    map[num] = i
  end
  [-1]
end

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum2(nums, target)
  nums.each_with_index do |num, i|
    complement = target - num

    for j in i + 1...nums.length
      return [i, j] if complement == nums[j]
    end
  end
  [-1]
end
